Support αiou loss

CCODING04 opened the pull request:
## Motivation

Support αiou loss

## Modification

add α-iou loss in `mmdet/models/losses/iou_loss.py`

## Use cases(TODO)

Take ATSS and YOLOX as example

## Checklist

- [x]  Implement αiou loss
- [x]  add unit tests to test your code
- [x]  docstring
- [ ]  Benchmark the performance(ATSS and YOLOX)
- [ ]  Prepare and release the models

CCODING04 committed:
add alphaIoU

CCODING04 committed:
fix typo with precommit

ZwwWayne commented on mmdet/models/losses/iou_loss.py:
```diff
@@ -237,6 +237,78 @@ def ciou_loss(pred, target, eps=1e-7):
     return loss
 
 
+@mmcv.jit(derivate=True, coderize=True)
```
this decorator is unnecessary

CCODING04 replied:
> this decorator is unnecessary

Okay, I will delete it in the next commit.

ZwwWayne commented on mmdet/models/losses/iou_loss.py:
```diff
@@ -237,6 +237,78 @@ def ciou_loss(pred, target, eps=1e-7):
+            with torch.no_grad():
+                alpha_ciou = (ious > 0.5).float() * v / (1 - ious + v)
+
+            # CIoU
```
This comment can be put at line 296 to improve readability of the code

CCODING04 replied:
> This comment can be put at line 296 to improve readability of the code

Okay, I will change it later. Thanks for your advice. :)
It's just a coarse implementation and will be improved it later.

CCODING04 committed:
fix loss calc error when mode iou

CCODING04 committed:
atss alphaiou test

jbwang1997 commented on mmdet/models/losses/iou_loss.py:
```diff
@@ -237,6 +237,79 @@ def ciou_loss(pred, target, eps=1e-7):
+    ag = (target[:, 2] - target[:, 0]) * (target[:, 3] - target[:, 1])
+    union = ap + ag - overlap + eps
+
+    # change conventional iou to alpha pow
+    ious = torch.pow(overlap / union + eps, alpha)
+
```
Is the eps here necessary? Seems direct `torch.pow(overlap / union + eps, alpha)` will not cause error.

CCODING04 replied:
> Is the eps here necessary? Seems direct `torch.pow(overlap / union + eps, alpha)` will not cause error.

You are right, it reminds me that maybe I add `eps` there is to make the program in consistent with others. I will check it later.

jbwang1997 commented on mmdet/models/losses/iou_loss.py:
```diff
@@ -237,6 +237,79 @@ def ciou_loss(pred, target, eps=1e-7):
+    ious = torch.pow(overlap / union + eps, alpha)
+
+    # calculate alpha-iou according mode
+    if mode == 'iou':
+        loss = 1 - ious
+        return loss
+    else:
+        # enclose area
+        enclose_x1y1 = torch.min(pred[:, :2], target[:, :2])
```
The `else` is unnecessary.

CCODING04 replied:
Sorry for not replying you in time, something has bothered me over the two weeks. 
I will resume to work from now, thanks.

jbwang1997 replied:
It's no problem. You can manage progress based on your schedule.

CCODING04 committed:
modified alpha-iou

CCODING04 committed:
update alpha_iou: add specified argument alpha

CCODING04 commented:
@jbwang1997 
Hi, I have trained YOLOX and ATSS with alpha-iou( giou and ciou) loss recently but got unexpected results (even worse 😭 ) . I will review my code on the weekend.

jbwang1997 commented:
It's Okay. Take your time. If you have any problem, you can also propose it in this pr.

CCODING04 committed:
check alpha-iou performance with original source code

CCODING04 committed:
update alpha-iou in mode iou: del log operation

CCODING04 committed:
Merge branch 'open-mmlab:master' into alpha_iou

CCODING04 committed:
test: tiny modified to reproduce expected results

CCODING04 commented:
> It's Okay. Take your time. If you have any problem, you can also propose it in this pr.

I have analyzed the results of my exps recently. Alpha-IOU could not improve the metrics in ATSS and YOLOX ( Also did exp with Faster-RCNN bud even worse than before) . According the author said, I just modify the IOU caculation in source code, and keep hyper params the same. 😢

jbwang1997 commented:
> > It's Okay. Take your time. If you have any problem, you can also propose it in this pr.
> 
> I have analyzed the results of my exps recently. Alpha-IOU could not improve the metrics in ATSS and YOLOX ( Also did exp with Faster-RCNN bud even worse than before) . According the author said, I just modify the IOU caculation in source code, and keep hyper params the same. 😢

It's okay, we will work together to figure out the problem.

I obverse the paper uses Faster R-CNN in mmdetection to compare the results of IoULoss and AlphaIoULoss. So, we can use [this config](https://github.com/open-mmlab/mmdetection/blob/master/configs/faster_rcnn/faster_rcnn_r50_fpn_iou_1x_coco.py) to align the results.

According to your Faster R-CNN config. I think maybe two points affect the results.
- The loss bbox of RPN needs to keep as original Faster RCNN.
- The loss weights of AlphaIoULoss should be 10 which is as the same as [this config](https://github.com/open-mmlab/mmdetection/blob/master/configs/faster_rcnn/faster_rcnn_r50_fpn_iou_1x_coco.py).

BTW, you can first try to set alpha as 1 to see if the results are equal to IoULoss. You can post your log and results here and we can discuss together.

CCODING04 committed:
delete dirty code & align the outputs of alphaiou calculation

CCODING04 committed:
align alphaiou outputs as gh repo clone CCODING04/AlphaIOUExpRec

CCODING04 committed:
add AlphaIoU

CCODING04 commented:
![AlphaIoU实验结果](https://user-images.githubusercontent.com/30226969/178409755-1515fa02-0d27-43a2-8176-65942cf4fe06.png)  
精度对齐放在了 https://github.com/CCODING04/AlphaIOUExpRec

CCODING04 commented:
> 

Thanks a lot, sorry for missing your response. Orz 
The config may mislead you, it was pushed by mistake.
In my local env, I tried different config in order to find the best performance. Modified RPN could not work well and I have already dropped it.
I am training yolox-tiny with alphaiou nowdays.
When it's over, I will post all my exp logs.
Thanks for your kindly advice!

CCODING04 commented:
That's all my exp logs of alpha-iou:
[faster_rcnn_r50_fpn_alphagiou_1x_coco_eps1e-9.log](https://github.com/open-mmlab/mmdetection/files/9129642/faster_rcnn_r50_fpn_alphagiou_1x_coco_eps1e-9.log)
[faster_rcnn_r50_fpn_alphaiou_1x_coco_eps1e-9.log](https://github.com/open-mmlab/mmdetection/files/9129643/faster_rcnn_r50_fpn_alphaiou_1x_coco_eps1e-9.log)
[faster_rcnn_r50_fpn_alphaiou_1x_coco_eps1e-6.log](https://github.com/open-mmlab/mmdetection/files/9129644/faster_rcnn_r50_fpn_alphaiou_1x_coco_eps1e-6.log)
[yolox_tiny_alphaiou_8x8_300e_coco_eps1e-9.log](https://github.com/open-mmlab/mmdetection/files/9129646/yolox_tiny_alphaiou_8x8_300e_coco_eps1e-9.log)
[atss_r50_fpn_alphaiou_1x_coco_eps1e-9.log](https://github.com/open-mmlab/mmdetection/files/9129648/atss_r50_fpn_alphaiou_1x_coco_eps1e-9.log)

jbwang1997 commented on mmdet/models/losses/iou_loss.py:
```diff
@@ -237,6 +237,88 @@ def ciou_loss(pred, target, eps=1e-7):
+        eps (float): Eps to avoid log(0).
+        alpha (int): adaptively reweighting the loss and gradient of IoU
+        mode (str): Options are "iou"(default), "giou", "diou" and "ciou"
+    Return:
+        Tensor: Loss tensor.
```
Need to add a blank line before Returns.

jbwang1997 commented on mmdet/models/losses/iou_loss.py:
```diff
@@ -472,3 +554,66 @@ def forward(self,
+                 loss_weight=1.0,
+                 mode='iou',
+                 alpha=3):
+        """AlphaIoULoss.
+
+        Computing the AlphaIoU loss between a set of predicted bboxes
+        and target bboxes.
+
+        Args:
+            eps (float): Eps to avoid log(0). We test serval exps and
+                        1e-9 performs better.
+            reduction (str): Options are "none", "mean" and "sum".
+            loss_weight (float): Weight of loss.
+            mode (str): the calculation method of IoU,
+                        options are "iou" (default), "giou", "diou" and "ciou".
+            alpha (int): adaptively reweighting the loss and gradient of IoU
+        """
```
The docstring of __init__ should between `class AlphaIoULoss` and `def __init__`.

jbwang1997 commented on mmdet/models/losses/iou_loss.py:
```diff
@@ -237,6 +237,88 @@ def ciou_loss(pred, target, eps=1e-7):
 
+@weighted_loss
+def alphaiou_loss(pred, target, eps=1e-9, alpha=3, mode='iou'):
+    r"""`Implementation of paper  Alpha-IoU: A Family of Power
+    Intersection over Union Losses for Bounding Box Regression
+    <https://arxiv.org/abs/2110.13675>` _.
+
```
Implementation of `Alpha-IoU <https://arxiv.org/abs/2110.13675>`_

jbwang1997 commented on mmdet/models/losses/iou_loss.py:
```diff
@@ -237,6 +237,88 @@ def ciou_loss(pred, target, eps=1e-7):
+              (torch.min(b1_y2, b2_y2) - torch.max(b1_y1, b2_y1)).clamp(0)
+
+    # union
+    w1, h1 = b1_x2 - b1_x1, b1_y2 - b1_y1 + eps
+    w2, h2 = b2_x2 - b2_x1, b2_y2 - b2_y1 + eps
+    union = w1 * h1 + w2 * h2 - overlap + eps
+
+    # change conventional iou to alpha pow
+    ious = torch.pow(overlap / union + eps, alpha)
```
We can only keep the eps at `union = w1 * h1 + w2 * h2 - overlap + eps` to avoid Nan. The other eps will not affect
 the final results.

CCODING04 replied:
> We can only keep the eps at `union = w1 * h1 + w2 * h2 - overlap + eps` to avoid Nan. The other eps will not affect the final results.

`eps` here is aim to  align the accuracy with the original implementation. Obviously it's redundant, emm, okay I drop it!

jbwang1997 commented:
Hello @CCODING04, thanks for your continued contributions. I left some comments above. Please remember to solve them. BTW, seems the results of Faster RCNN don't reach the results in the paper, Have you tried to optimize hyperparameters like weight and alpha?

CCODING04 committed:
fix docstring and indent problem

CCODING04 commented:
> Hello @CCODING04, thanks for your continued contributions. I left some comments above. Please remember to solve them. BTW, seems the results of Faster RCNN don't reach the results in the paper, Have you tried to optimize hyperparameters like weight and alpha?

I tried, and the accuracy seems a littel better than default settings with alpha=3, but far from the results of paper.
